lib/aws/json.ex

```diff
@@ -0,0 +1,23 @@
+defmodule AWS.JSON do
+  @moduledoc """
+  Specifies the behaviour of a JSON parser.
+
+  You can switch the default JSON parser which uses Poison underneath
+  by defining a different implementation by setting the `:json_module`
+  configuration in AWS.Client:
+
+     client = %AWS.Client{json_module: MyCustomJSONParser}
+     Aws.SNS.publish(client, %{})
+
+  """
+
+  @doc """
+  Encodes a map into JSON string. Raises in case of errors.
+  """
+  @callback encode!(input :: map(), options :: keyword()) :: String.t()
+
+  @doc """
+  Decodes a JSON string into a map. Raises in case of errors.
+  """
+  @callback decode!(input :: String.t(), options :: keyword()) :: map()
```

Comment:
I think the description of a JSON "type" is more than a `map()`: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/JSON

I saw that Poison describe it as `@type t :: nil | true | false | list | float | integer | String.t() | map`, and Jason uses only `term` to describe it. We can choose one of them an replace the `map()` here by it. WDYT?

Reply:
I wrote `map()` because this lib expects a map everywhere. Do you think `term` would be better?

Reply:
You have a point. I didn't see the usage of anything that isn't a map, so let's stick with if for now :smile: